core/src/upgrade.rs

```diff
@@ -152,3 +154,22 @@ pub trait OutboundConnectionUpgrade<T>: UpgradeInfo {
     /// The `info` is the identifier of the protocol, as produced by `protocol_info`.
     fn upgrade_outbound(self, socket: T, info: Self::Info) -> Self::Future;
 }
+
+/// Possible security upgrade on a connection
+pub trait SecurityUpgrade<T>: UpgradeInfo {
+    /// Output after the upgrade has been successfully negotiated and the handshake performed.
+    type Output;
+    /// Possible error during the handshake.
+    type Error;
+    /// Future that performs the handshake with the remote.
+    type Future: Future<Output = Result<Self::Output, Self::Error>>;
```

Comment:
```suggestion
    type Future: Future<Output = Result<(PeerId, Self::Output), Self::Error>>;
```

With a specialised upgrade, we can fix the `PeerId` here.

Reply:
That makes sense, thanks.